src/core/TsApi.py

```diff
@@ -4,144 +4,332 @@
+        logging.info("Waiting for jobs to finish...")
+        self.executor.shutdown(wait=True, cancel_futures=True)
+        
+        # Final database sync
+        self.database.sync()
         logging.info("TsAPI stopped!")
         sys.exit(0)
-
+    
     def add_to_queue(self, priority: int, module_entry: Default.Entry) -> None:
-        """
-        Fügt einen Job zur Warteschlange hinzu.
-
-        :param priority: Die Priorität des Jobs.
-        :param module_entry: Der Eintrag, der zur Warteschlange hinzugefügt
-        werden soll.
-        """
-        logging.info(f"Adding job with id {module_entry.uid} to queue.")
+        """Add a job to the queue"""
+        logging.info(f"Adding job {module_entry.uid} to queue with priority {priority}")
         try:
+            # Add to database
             self.database.add_job(module_entry)
-            self.database.change_job_entry(module_entry.uid, "status", 0)
-            self.database.queue.put((priority, module_entry))
+            self.database.update_job_status(module_entry.uid, JobStatus.QUEUED)
+            self.database.add_to_queue(module_entry.uid, priority)
+            
+            # Add to in-memory queue
+            self.job_queue.put((priority, module_entry))
         except Exception as e:
             logging.error(f"Error adding job {module_entry.uid} to queue: {e}")
-
-    # Track running jobs
-    def register_job(self, entry: Default.Entry) -> Transcriber:
+    
+    def cancel_job(self, job_id: str, requeue: bool = False) -> bool:
         """
-        Registriert einen laufenden Job und erstellt einen Transcriber.
-
-        :param entry: Der Eintrag des Jobs.
-        :return: Der vorbereitete Transcriber.
+        Cancel a running or queued job
+        
+        Args:
+            job_id: The job UID to cancel
+            requeue: If True, requeue the job for later processing
+        
+        Returns:
+            True if job was canceled, False otherwise
         """
-        logging.info(f"Starting job with id {entry.uid}.")
-        trans = Transcriber(self, entry)
-        return trans
-
+        with self.job_lock:
+            # Check if job is running
+            if job_id in self.job_futures:
+                future = self.job_futures[job_id]
+                
+                # Try to cancel if not started
+                if future.cancel():
+                    logging.info(f"Canceled job {job_id} before it started")
+                    self._cleanup_job(job_id, requeue=requeue)
+                    return True
+                
+                # If running, signal cancellation
+                if job_id in self.cancel_events:
+                    logging.info(f"Signaling cancellation for running job {job_id}")
+                    self.cancel_events[job_id].set()
+                    
+                    if requeue:
+                        # Mark for requeue
+                        if job_id in self.module_entrys:
```

Comment:
```suggestion
                        if job_id in self.database.module_entrys:
```
In the initialisation of the class we don't have a variable `self.module_entrys` which means it is probably a buggy reference. Everywhere else we refer as `self.database.module_entrys` as a part of Database object.